File: 1000/1654.cpp
#include<iostream>
#include<algorithm>
using namespace std;
#define ll long long

int n, k;
int line[10001];

void input() {
	cin >> k >> n;
	for (int i = 0; i < k; ++i) cin >> line[i];
}

void solution() {
	input();
	sort(line, line + k);
	ll low = 1, high = line[k - 1];
	int answer = 0;
	while (low <= high) {
		ll mid = (low + high) >> 1;
		int cnt = 0;
		for (int i = 0; i < k; ++i) {
			cnt += line[i] / mid;
		}
		if (cnt >= n) {
			if (mid > answer) answer = mid;
			low = mid + 1;
		}
		else {
			high = mid - 1;
		}
	}
	cout << answer;
}

int main() {
	cin.tie(0), cout.tie(0), ios_base::sync_with_stdio(0);
	solution();
	return 0;
}

File: 1000/1654.java
import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int n, k;
    static long l, r, answer = 0;
    static int[] line;

    static void input() throws IOException {
        st = new StringTokenizer(br.readLine());
        k = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());
        line = new int[k];
        l = 1;
        for (int i = 0; i < k; ++i) {
            line[i] = Integer.parseInt(br.readLine());
            if(line[i] > r) r = line[i];
        }
    }

    static int getMaxLength(long len) {
        int cnt = 0;
        for (int i = 0; i < k; ++i) {
            cnt += line[i] / len;
        }
        return cnt;
    }

    static void binarySearch() {
        while (l <= r) {
            long mid = (l + r) >> 1;
            int cnt = getMaxLength(mid);
            if (cnt < n) {
                r = mid - 1;
            }
            else{
                answer = mid;
                l = mid + 1;
            }
        }
    }

    static void solution() {
        Arrays.sort(line);
        binarySearch();
        System.out.println(answer);
    }

    public static void main(String[] args) throws IOException {
        input();
        solution();
    }
}
